src/antlr/QueryStructureVisitor.ts

```diff
@@ -0,0 +1,522 @@
+
+export abstract class QueryStructureVisitor<Result> extends AbstractParseTreeVisitor<Result>
+  implements SqlBaseVisitor<Result> {
+  private relationSeq = 0;
+
+  public readonly relationsStack: Array<QueryRelation> = [];
+
+  protected currentRelation = new QueryRelation("result_0");
+
+  constructor(
+    public getTable: (
+      table: TablePrimary
+    ) => { table: { id: string; data: unknown }; columns: { id: string; data: unknown }[] } | undefined
+  ) {
+    super();
+  }
+
+  getNextRelationId(): string {
+    this.relationSeq++;
+    return "result_" + this.relationSeq;
+  }
+
+  rangeFromContext(ctx: ParserRuleContext): Range {
+    const stop = ctx.stop ?? ctx.start;
+    return {
+      startLine: ctx.start.line,
+      endLine: stop.line,
+      startColumn: ctx.start.charPositionInLine,
+      endColumn: stop.charPositionInLine + (stop.stopIndex - stop.startIndex + 1)
+    };
+  }
+
+  /**
+   *  Extracts table and column names from PrimaryExpressionContext (if possible).
+   */
+
+  protected extractTableAndColumn(
+    ctx: PrimaryExpressionContext
+  ): { table?: QuotableIdentifier; column: QuotableIdentifier } | undefined {
+    if (ctx instanceof ColumnReferenceContext) {
+      return { column: common.stripQuote(ctx.identifier()) };
+    } else if (ctx instanceof DereferenceContext) {
+      const primary = ctx.primaryExpression();
+      if (primary instanceof ColumnReferenceContext) {
+        return {
+          table: common.stripQuote(primary.identifier()),
+          column: common.stripQuote(ctx.identifier())
+        };
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   *  Derives column name from expression if possible.
+   */
+
+  protected deriveColumnName(ctx: ExpressionContext): string | undefined {
+    const boolExpr = ctx.booleanExpression();
+    if (boolExpr instanceof PredicatedContext) {
+      const valExpr = boolExpr.valueExpression();
+      if (valExpr instanceof ValueExpressionDefaultContext) {
+        const tableCol = this.extractTableAndColumn(valExpr.primaryExpression());
+        if (tableCol !== undefined) {
+          return tableCol.column.name;
+        }
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   *  Determines whether expression is a star.
+   */
+
+  private isStar(ctx: ExpressionContext): StarContext | undefined {
+    const boolExpr = ctx.booleanExpression();
+    if (boolExpr instanceof PredicatedContext) {
+      const valExpr = boolExpr.valueExpression();
+      if (valExpr instanceof ValueExpressionDefaultContext) {
+        const primaryExpr = valExpr.primaryExpression();
+        if (primaryExpr instanceof StarContext) {
+          return primaryExpr;
+        }
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   * Called when relation is ready.
+   * @param _relation
+   * @param _alias
+   * @returns
+   */
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  onRelation(_relation: TableRelation | QueryRelation, _alias?: string): void {
```

Comment:
Two questions:
1. Do we need the underscore even with the linting directive?
2. Since the class is `abstract` can we omit the implementation here?

Reply:
1. Somehow linting directive is not enough. But I removed linting directive in favor of underscores.
2. The idea was that not all subclasses want to implement these methods.

Reply:
Starting by underscore is the way to say that the variable is not used 😉 No problem with this, this is the convention apply in this repository